### PR TITLE
OpenGL renderer updates

### PR DESCRIPTION
**Pull request by eulertour**
## Changelog / Overview

Improvements to OpenGL infrastructure

## Motivation

Various improvements to OpenGL rendering and quality of life additions. 

## Explanation for Changes

* Adds model matrices to all OpenGLVMobjects
* Improved performance on vectorized mobject shaders
* Added updaters that are part of the scene rather than a mobject

[Documentation Reference](https://manimce--####.org.readthedocs.build)

## Testing Status

Stroke shaders can be tested with any scene with vectorized mobjects and the model matrices can be tested by moving a mobject with e.g. `mob.model_matrix = translation_matrix(x=2) @ mob.model_matrix`.

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [ ] My new functions/classes either have a docstring or are private
- [ ] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

File: manim/renderer/opengl_renderer.py [hunk 1]
```diff
@@ -43,15 +42,26 @@ def __init__(
         focal_distance=2,
         light_source_position=[-10, 10, 10],
         orthographic=False,
+        minimum_polar_angle=-PI / 2,
+        maximum_polar_angle=PI / 2,
```
Comment (hydrobeam) [inline manim/renderer/opengl_renderer.py:46]: I'm not sure why, but these limits don't do anything when using 3b1b's shaders, only the projection ones.

**Review by hydrobeam — COMMENTED** [review]
I'm not sure why, but using the projection shaders makes the mobject look more... pixelated?
Here's an animation usuing the projection shaders:
```py
manim mkt.py -f huh --renderer=opengl --write_to_movie --disable_caching

class huh(Scene):
    def construct(self):
        a = Star()
        self.play(Write(a))
        self.wait()
        # self.interactive_embed()
```

https://user-images.githubusercontent.com/70682032/126232457-8818aeb3-6cb2-4f93-aa34-c64e935f7ea7.mp4


Using 3b1b's shaders:

https://user-images.githubusercontent.com/70682032/126232537-579d0343-1c31-458f-b997-7fd62bb2979d.mp4

I tested using `main` and it doesn't seem to have been introduced in this PR, but I thought it was worth mentioning.


Also ThreeDMobjectTest has some flickering (not introduced by this PR too):

https://user-images.githubusercontent.com/70682032/126234780-23a32e15-a11f-4e94-b79c-7647020b9c8e.mp4



also, do you have any sample code to test model matrices?

File: manim/renderer/opengl_renderer.py (hunk 1 above)
Comment (eulertour) [reply manim/renderer/opengl_renderer.py:46]: Pretty much none of the OpenGL features I wrote will work with those shaders. They don't follow any of the same standards that pretty much every other shader library uses.

**Review by hydrobeam — APPROVED** [review]
I looked through it and the manim-specific code looks good 👍. Also the performance has notciably improved 🥳!

Although I guess I have two questions:
> Adds model matrices to all OpenGLVMobjects

What's the advantage of using these over regular translations via `mob.shift()`?

> Added updaters that are part of the scene rather than a mobject

What does this change allow for/ fix?

**Comment by eulertour** [comment]
> What's the advantage of using these over regular translations via mob.shift()?

They abstract away a lot of math that you'd otherwise have to calculate manually, for example mobjects that are fixed relative to the camera.

> What does this change allow for/ fix?

Sometimes you want to run some code on each frame but it isn't logically related to any individual mobject, so this is to allow adding it to the scene instead.

**Pull request merged**